File: src/main/java/org/jcodec/containers/mp4/boxes/TimeToSampleBox.java
package org.jcodec.containers.mp4.boxes;

import java.nio.ByteBuffer;

import org.jcodec.containers.mp4.boxes.Box.AtomField;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * A box containing sample presentation time information
 * 
 * @author The JCodec project
 * 
 */
public class TimeToSampleBox extends FullBox {

    public TimeToSampleBox(Header atom) {
        super(atom);
    }

    public static class TimeToSampleEntry {
        int sampleCount;
        int sampleDuration;

        public TimeToSampleEntry(int sampleCount, int sampleDuration) {
            this.sampleCount = sampleCount;
            this.sampleDuration = sampleDuration;
        }

        @AtomField(idx=0)
        public int getSampleCount() {
            return sampleCount;
        }

        @AtomField(idx=1)
        public int getSampleDuration() {
            return sampleDuration;
        }

        public void setSampleDuration(int sampleDuration) {
            this.sampleDuration = sampleDuration;
        }

        public void setSampleCount(int sampleCount) {
            this.sampleCount = sampleCount;
        }

        public long getSegmentDuration() {
            return sampleCount * sampleDuration;
        }
    }

    public static String fourcc() {
        return "stts";
    }

    public static TimeToSampleBox createTimeToSampleBox(TimeToSampleEntry[] timeToSamples) {
        TimeToSampleBox box = new TimeToSampleBox(new Header(fourcc()));
        box.entries = timeToSamples;
        return box;
    }

    private TimeToSampleEntry[] entries;

    public void parse(ByteBuffer input) {
        super.parse(input);
        int foo = input.getInt();
        entries = new TimeToSampleEntry[foo];
        for (int i = 0; i < foo; i++) {
            entries[i] = new TimeToSampleEntry(input.getInt(), input.getInt());
        }
    }

    @AtomField(idx=0)
    public TimeToSampleEntry[] getEntries() {
        return entries;
    }

    @Override
    public void doWrite(ByteBuffer out) {
        super.doWrite(out);
        out.putInt(entries.length);
        for (int i = 0; i < entries.length; i++) {
            TimeToSampleEntry timeToSampleEntry = entries[i];
            out.putInt(timeToSampleEntry.getSampleCount());
            out.putInt(timeToSampleEntry.getSampleDuration());
        }
    }
    
    @Override
    public int estimateSize() {
        return 16 + entries.length * 8;
    }

    public void setEntries(TimeToSampleEntry[] entries) {
        this.entries = entries;
    }
}

File: src/main/java/org/jcodec/containers/mp4/boxes/TimeToSampleBox.kt
package org.jcodec.containers.mp4.boxes

import java.nio.ByteBuffer

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * A box containing sample presentation time information
 *
 * @author The JCodec project
 */
class TimeToSampleBox(atom: Header) : FullBox(atom) {
    class TimeToSampleEntry(var sampleCount: Int, var sampleDuration: Int) {

        val segmentDuration: Long
            get() = (sampleCount * sampleDuration).toLong()

    }

    private var entries: Array<TimeToSampleEntry?> = emptyArray()
    override fun parse(input: ByteBuffer) {
        super.parse(input)
        val foo = input.int
        entries = arrayOfNulls(foo)
        for (i in 0 until foo) {
            entries[i] = TimeToSampleEntry(input.int, input.int)
        }
    }

    fun getEntries(): Array<TimeToSampleEntry?> {
        return entries
    }

    public override fun doWrite(out: ByteBuffer) {
        super.doWrite(out)
        out.putInt(entries.size)
        for (i in entries.indices) {
            val timeToSampleEntry = entries[i]
            out.putInt(timeToSampleEntry!!.sampleCount)
            out.putInt(timeToSampleEntry.sampleDuration)
        }
    }

    override fun estimateSize(): Int {
        return 16 + entries.size * 8
    }

    fun setEntries(entries: Array<TimeToSampleEntry?>) {
        this.entries = entries
    }

    companion object {
        @JvmStatic
        fun fourcc(): String {
            return "stts"
        }

        @JvmStatic
        fun createTimeToSampleBox(timeToSamples: Array<TimeToSampleEntry?>): TimeToSampleBox {
            val box = TimeToSampleBox(Header(fourcc()))
            box.entries = timeToSamples
            return box
        }
    }
}